packages/hardhat/contracts/periphery/lib/ControllerHelperUtil.sol

```diff
@@ -71,14 +79,41 @@ library ControllerHelperUtil {
      * @param _isWethToken0 bool variable indicate if Weth token is token0 in Uniswap v3 weth/wPowerPerp pool
      * @return _vaultId and tokenId
      */
-    function mintAndLp(address _controller, address _nonfungiblePositionManager, address _wPowerPerp, address _wPowerPerpPool, ControllerHelperDataType.MintAndLpParams calldata _mintAndLpParams, bool _isWethToken0) public returns (uint256, uint256) {
+    function mintAndLp(address _controller, address _nonfungiblePositionManager, address _wPowerPerp, address _wPowerPerpPool, address _weth, address _oracle, ControllerHelperDataType.MintAndLpParams calldata _mintAndLpParams, bool _isWethToken0) public returns (uint256, uint256) {
+        IWETH9(_weth).withdraw(_mintAndLpParams.collateralToDeposit);
+
+        uint256 amount0Desired; 
+        uint256 amount1Desired;
+        {
+            int24 timeWeightedAverageTick = IOracle(_oracle).getTimeWeightedAverageTickSafe(_wPowerPerpPool, TWAP_PERIOD);
+            uint160 sqrtRatioX96 = TickMathExternal.getSqrtRatioAtTick(timeWeightedAverageTick);
+            uint160 sqrtRatioAX96 = TickMathExternal.getSqrtRatioAtTick(_mintAndLpParams.lowerTick);
+            uint160 sqrtRatioBX96 = TickMathExternal.getSqrtRatioAtTick(_mintAndLpParams.upperTick);
+            (amount0Desired, amount1Desired) = _isWethToken0 ? (_mintAndLpParams.collateralToLp, _mintAndLpParams.wPowerPerpAmount) : (_mintAndLpParams.wPowerPerpAmount, _mintAndLpParams.collateralToLp);
+            uint128 maxLiquidity = LiquidityAmounts.getLiquidityForAmounts(uint160(timeWeightedAverageTick), sqrtRatioAX96, sqrtRatioBX96, amount0Desired, amount1Desired);
```

Comment:
Some of tests are failing because ```timeWeightedAverageTick``` is used as a param instead of ```sqrtRatioX96``` here.  

The other tests are failing as the ticks are not correct (I'm pretty sure).  We can just use 0 as the dividing tick for all oSQTH or all ETH, a tick of 0 corresponds to a price of 1.0001^0=1, where a tick of 1000 corresponds to a price of 1.0001^1000 = 1.1, etc.

I revamped the process in the integration test file to calculate all of this correctly as it was a problem there as well.

Reply:
Code for how I fixed the tests for LP of specified composition: https://github.com/opynfinance/squeeth-monorepo/blob/main/packages/hardhat/test/integration-tests/periphery/controller-helper.ts#L1471